<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: try an expression, returning the error stack</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for tryStack {berryFunctions}"><tr><td>tryStack {berryFunctions}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>try an expression, returning the error stack</h2>

<h3>Description</h3>

<p>As in <code><a href="../../base/html/try.html">try</a></code>, the result of an expression if it works.
If it fails, execution is not halted, but an invisible try-error class object is
returned and (unless silent=TRUE) a message <code><a href="../../base/html/cat.html">cat</a>ted</code> to the console.<br />
Unlike <code><a href="../../base/html/try.html">try</a></code>, <code>tryStack</code> also returns the calling stack to
trace errors and warnings and ease debugging.
</p>


<h3>Usage</h3>

<pre>
tryStack(
  expr,
  silent = FALSE,
  warn = TRUE,
  short = TRUE,
  file = "",
  removetry = TRUE,
  skip = NULL
)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>expr</code></td>
<td>
<p>Expression to try, potentially wrapped in curly braces if
spanning several commands.</p>
</td></tr>
<tr valign="top"><td><code>silent</code></td>
<td>
<p>Logical: Should printing of error message + stack be suppressed?
Does not affect warnings and messages. DEFAULT: FALSE</p>
</td></tr>
<tr valign="top"><td><code>warn</code></td>
<td>
<p>Logical: trace <code><a href="../../base/html/warning.html">warning</a>s</code> and <code><a href="../../base/html/message.html">message</a>s</code> also?
They are still handled like regular warnings / messages unless
<code>file !=""</code>, when they are catted into that file. DEFAULT: TRUE</p>
</td></tr>
<tr valign="top"><td><code>short</code></td>
<td>
<p>Logical: should trace be abbreviated to upper -&gt; middle -&gt; lower?
If NA, it is set to TRUE for warnings and messages, FALSE for errors.
DEFAULT: TRUE</p>
</td></tr>
<tr valign="top"><td><code>file</code></td>
<td>
<p>File name passed to <code><a href="../../base/html/cat.html">cat</a></code>.
If given, Errors will be appended to the file after two empty lines.
if <code>warn=T</code> and file!=&quot;&quot;, warnings and messages will not be shown,
but also appended to the file.
This is useful in lapply simulation runs.
DEFAULT: &quot;&quot; (catted to the console)</p>
</td></tr>
<tr valign="top"><td><code>removetry</code></td>
<td>
<p>Logical: should all stack entries matching typical tryCatch
expressions be removed? Unless the call contains customized
<code><a href="../../base/html/conditions.html">tryCatch</a></code> code, this can be left to the DEFAULT: TRUE</p>
</td></tr>
<tr valign="top"><td><code>skip</code></td>
<td>
<p>Character string(s) to be removed from the stack.
e.g. &quot;eval(expr, p)&quot;. Use short=F to find exact matches.
DEFAULT: NULL</p>
</td></tr>
</table>


<h3>Value</h3>

<p>Value of <code>expr</code> if evaluated successfully. If not, an invisible
object of class &quot;try-error&quot; as in <code><a href="../../base/html/try.html">try</a></code> with the stack in <code>object[2]</code>.
For nested tryStack calls, <code>object[3], object[4]</code> etc. will contain &quot;&ndash; empty error stack &ndash;&quot;
</p>


<h3>Author(s)</h3>

<p>Berry Boessenkool, <a href="mailto:berry-b@gmx.de">berry-b@gmx.de</a>, Nov 2016
</p>


<h3>See Also</h3>

<p><code><a href="../../base/html/try.html">try</a></code>, <code><a href="traceCall.html">traceCall</a></code>,
<a href="http://r.789695.n4.nabble.com/Stack-trace-td4021537.html">http://r.789695.n4.nabble.com/Stack-trace-td4021537.html</a>,
<a href="http://stackoverflow.com/questions/15282471/get-stack-trace-on-trycatched-error-in-r">http://stackoverflow.com/questions/15282471/get-stack-trace-on-trycatched-error-in-r</a>,
<a href="http://stackoverflow.com/questions/1975110/printing-stack-trace-and-continuing-after-error-occurs-in-r">http://stackoverflow.com/questions/1975110/printing-stack-trace-and-continuing-after-error-occurs-in-r</a>,
<a href="http://stackoverflow.com/questions/16879821/save-traceback-on-error-using-trycatch">http://stackoverflow.com/questions/16879821/save-traceback-on-error-using-trycatch</a>
</p>


<h3>Examples</h3>

<pre>

# Functions -----

lower &lt;- function(a) {message("fake message, a = ", a); a+10}
middle &lt;- function(b) {plot(b, main=b) ; warning("fake warning, b = ", b); lower(b) }
upper &lt;- function(c) {cat("printing c:", c, "\n") ; middle(c)}
d &lt;- upper(42)
d
rm(d)


# Classical error management with try -----

is.error( d &lt;- upper("42"), TRUE, TRUE) # error, no d creation
traceback()                     # calling stack, but only in interactive mode

d &lt;- try(upper("42"), silent=TRUE)      # d created
cat(d)                                  # with error message, but no traceback
inherits(d, "try-error")                # use for coding


# way cooler with tryStack -----

d &lt;- tryStack(upper("42") ) # like try, but with traceback, even for warnings
cat(d)
d &lt;- tryStack(upper("42"), silent=TRUE, warn=0) # don't trace warnings
d &lt;- tryStack(upper("42"), short=FALSE)

tryStack(upper(42)) # returns normal output, but warnings are easier to debug
# Note: you can also set options(showWarnCalls=TRUE)

stopifnot(inherits(d, "try-error"))
stopifnot(tryStack(upper(42))==52)

## Not run:  ## file writing not wanted by CRAN checks
d &lt;- tryStack(upper("42"), silent=TRUE, file="log.txt")
openFile("log.txt")
unlink("log.txt")

## End(Not run)

op &lt;- options(warn=2)
d &lt;- try(upper("42") )
cat(d)
d &lt;- tryStack(upper("42") )
d &lt;- tryStack(upper("42"), warn=FALSE)
cat(d)
options(op) ; rm(op)

# Nested calls -----

f &lt;- function(k) tryStack(upper(k), silent=TRUE)
d &lt;- f(42)                 ; cat("-----\n", d, "\n-----\n") ; rm(d)
d &lt;- f("42")               ; cat("-----\n", d, "\n-----\n") ; rm(d)
d &lt;- tryStack(f(4) )       ; cat("-----\n", d, "\n-----\n") ; rm(d)
# warnings in nested calls are printed twice, unless warn=0
d &lt;- tryStack(f(4), warn=0) # could also be set within 'f'

d &lt;- tryStack(f("4"))      ; cat("-----\n", d, "\n-----\n")
d[1:3] ; rm(d)
# empty stack at begin - because of tryStack in f, no real error happened in f


# Other tests -----

cat( tryStack(upper("42")) )
f &lt;- function(k) tryStack(stop("oh oh"))
d &lt;- f(42) ; cat("-----\n", d, "\n-----\n") ; rm(d) # level 4 not helpful, but OK

# stuff with base::try
f &lt;- function(k) try(upper(k), silent=TRUE)
d &lt;- f(42)     ; cat("-----\n", d, "\n-----\n") ; rm(d)
d &lt;- f("42")   ; cat("-----\n", d, "\n-----\n") ; rm(d) # regular try output

f2 &lt;- function(k) tryStack(f(k), warn=0, silent=TRUE)
d &lt;- f2(42)    ; cat("-----\n", d, "\n-----\n") ; rm(d)
d &lt;- f2("42")  ; cat("-----\n", d, "\n-----\n") ; rm(d) # try -&gt; no error.
# -&gt; Use tryCatch and you can nest those calls. note that d gets longer.


</pre>

<hr /><div style="text-align: center;">[Package <em>berryFunctions</em> version 1.19.0 <a href="00Index.html">Index</a>]</div>
</body></html>
